### PR TITLE
Copy question link from dashboard

### LINKED ISSUE #41070: Copy question link from dashboard
**Context**
- product doc: https://www.notion.so/metabase/Copy-question-link-from-dashboard-702fe6ec9c874f12bb2cda19dbbf6f47

Allow copying the question link address from the dashboard

### PR DESCRIPTION
**Pull request by kamilmielnik**
Closes #41070

Original PR (#42892) was reverted (#43499).

It's recommended to review it with hidden whitespace changes due to deep nesting in `Visualization.jsx`.

### Description

Makes titles of most dashboard charts actual HTML anchors.
Thanks to this users can right-click the title to copy URL, middle-click it to open it in a new tab, drag & drop it to browser's address bar, etc.

### Performance

_"Scripting" times, in ms, average value shown first, 3 respective measurements in parens._

##### Prerequisites
1. Create a question Orders > Limit 1
2. Create a dashboard
3. Add **100 cards** with the question to the dashboard
4. Add text filter to the dashboard and connect it to all cards to the email field
5. Save dashboard

#### Scenario - Full dashboard load from scratch

Measurements:
  - `master`: 4354 (4381, 4345, 4337)
  - This PR: 4378 (4388, 4393, 4355)
    - `getNewCardUrl`: 194 (217, 174, 192)

The overhead is negligible in this case. 

#### Scenario - Changing dashboard parameters

1. Open up the text filter
2. Type a string
3. Start measuring performance
4. Hit "Add filter"
5. Wait for all cards to load
6. Remove filter
7. Wait for all cards to load
8. Stop measuring performance

Measurements:
  - `master`: 5595 (5611, 5580, 5594)
  - This PR:  6241 (6190, 6329, 6206)

The overhead is ~11%.

### How to verify

1. Create a dashboard
2. Add various visualizations to it
3. Right-click visualization title 

Context menu should open and you should be able to copy the URL of the link.

### Demo

https://github.com/metabase/metabase/assets/6830683/b31a0664-7367-4fd0-b6bc-9ce5a21eed1c

### CODE REVIEW

File: frontend/src/metabase/visualizations/components/Visualization/Visualization.jsx [hunk 1]
```diff
@@ -151,7 +153,14 @@ class Visualization extends PureComponent {
     const computedSettings = !this.isLoading(series)
       ? getComputedSettingsForSeries(series)
       : {};
+
+    const nextCard = series[0].card;
+    const previousCard =
+      newProps.rawSeries.find(series => series.card.id === nextCard?.id)
+        ?.card ?? newProps.rawSeries[0].card;
+
```
Comment (kamilmielnik) [inline frontend/src/metabase/visualizations/components/Visualization/Visualization.jsx:161]: This code is the same as in `handleOnChangeCardAndRun`

File: frontend/src/metabase/visualizations/components/Visualization/Visualization.jsx [hunk 2]
```diff
@@ -151,7 +153,14 @@ class Visualization extends PureComponent {
     const computedSettings = !this.isLoading(series)
       ? getComputedSettingsForSeries(series)
       : {};
+
+    const nextCard = series[0].card;
+    const previousCard =
+      newProps.rawSeries.find(series => series.card.id === nextCard?.id)
+        ?.card ?? newProps.rawSeries[0].card;
```
Comment (kamilmielnik) [inline frontend/src/metabase/visualizations/components/Visualization/Visualization.jsx:160]: This code is the same as in `handleOnChangeCardAndRun`

File: frontend/src/metabase/visualizations/components/Visualization/Visualization.jsx [hunk 3]
```diff
@@ -151,7 +153,14 @@ class Visualization extends PureComponent {
...
+
     this.setState({
+      href: newProps.getHref?.({ nextCard, previousCard }),
```
Comment (kamilmielnik) [inline frontend/src/metabase/visualizations/components/Visualization/Visualization.jsx:163]: Computing `href` here is optimal performance-wise

File: frontend/src/metabase/dashboard/components/DashCard/DashCard.tsx [hunk 4]
```diff
@@ -254,26 +259,36 @@ function DashCardInner({
...
-        dashcard,
-        objectId,
-      });
-    };
+  const metadata = useSelector(getMetadata);
+  const parameterValues = useSelector(getParameterValues);
 
-    return handler;
-  }, [dashcard, navigateToNewCardFromDashboard]);
+  const href = useMemo(() => {
```
Comment (kamilmielnik) [inline frontend/src/metabase/dashboard/components/DashCard/DashCard.tsx:278]: I brought it closer to the usage.
It turns out `nextCard` will always be the same as `previousCard` when clicking on title.

**Review by ranquild — APPROVED** [review]

**Pull request merged**